### PR TITLE
Add 51% attack contract

### PR DESCRIPTION
**Pull request by facuspagnuolo**

### CODE REVIEW

File: mainnet/X-51-attack/TheLobbyGuy.sol [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+        uint256 currentBalance = _token.balanceOf(address(this));
+        require(currentBalance > 0, ERROR_NO_BALANCE_TO_DISTRIBUTE);
+
+        (Voting.VoterState state,) = voting.getCastVote(voteId, msg.sender);
+        require(state == Voting.VoterState.Yea, ERROR_VOTER_DID_NOT_SUPPORT);
+
+        MiniMeToken votingToken = voting.token();
+        uint256 voterStake = votingToken.balanceOfAt(msg.sender, snapshotBlock);
+        uint256 share = yeas / voterStake;
```
Comment (sohkai) [inline mainnet/X-51-attack/TheLobbyGuy.sol]:
```suggestion
        uint256 share = voterStake / yeas;
```

Wouldn't this be the other way around?

File: mainnet/X-51-attack/TheLobbyGuy.sol [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        require(_token.transfer(msg.sender, share), ERROR_TOKEN_TRANSFER_FAILED);
+    }
```
Comment (sohkai) [inline mainnet/X-51-attack/TheLobbyGuy.sol]: Perhaps we should add a time period for withdrawals, after which the owner can withdraw any leftover funds not claimed.

**Review by sohkai — APPROVED** [review]

File: mainnet/X-51-attack/TheLobbyGuy.sol (hunk 1 above)
Comment (facuspagnuolo) [reply mainnet/X-51-attack/TheLobbyGuy.sol]: Haha I think I was a bit sleepy. I should actually be `stake * balance / yeas`

File: mainnet/X-51-attack/TheLobbyGuy.sol [hunk 3]
```diff
@@ -44,7 +98,16 @@ contract TheLobbyGuy {
 
         MiniMeToken votingToken = voting.token();
         uint256 voterStake = votingToken.balanceOfAt(msg.sender, snapshotBlock);
-        uint256 share = yeas / voterStake;
+
+        uint256 balance = _token.balanceOf(address(this));
```
Comment (sohkai) [inline mainnet/X-51-attack/TheLobbyGuy.sol]: This is already provided in `currentBalance` :)

File: mainnet/X-51-attack/TheLobbyGuy.sol [hunk 4]
```diff
@@ -19,20 +47,46 @@ interface Voting {
...
+    string private constant ERROR_SENDER_NOT_OWNER = "TLG_SENDER_NOT_OWNER";
+    string private constant ERROR_CANNOT_RECOVER_FUNDS = "TLG_CANNOT_RECOVER_FUNDS";
     string private constant ERROR_VOTE_WAS_NOT_EXECUTED = "TLG_VOTE_WAS_NOT_EXECUTED";
     string private constant ERROR_VOTER_DID_NOT_SUPPORT = "TLG_VOTER_DID_NOT_SUPPORT";
     string private constant ERROR_TOKEN_TRANSFER_FAILED = "TLG_TOKEN_TRANSFER_FAILED";
     string private constant ERROR_NO_BALANCE_TO_DISTRIBUTE = "TLG_NO_BALANCE_TO_DISTRIBUTE";
+    string private constant ERROR_VOTER_HAS_ALREADY_CLAIMED = "TLG_VOTER_HAS_ALREADY_CLAIMED";
+
+    address public owner;
+    uint256 public withdrawsEndDate;
```
Comment (sohkai) [inline mainnet/X-51-attack/TheLobbyGuy.sol:63]: I don't think three's a way to set this at the moment :)

File: mainnet/X-51-attack/TheLobbyGuy.sol [hunk 5]
```diff
@@ -19,20 +47,46 @@ interface Voting {
...
 
     Voting public voting;
     uint256 public voteId;
+    mapping (address => bool) hasClaimed;
```
Comment (sohkai) [inline mainnet/X-51-attack/TheLobbyGuy.sol]: Haha, good catch!

**Review by sohkai — APPROVED** [review]

File: mainnet/X-51-attack/TheLobbyGuy.sol (hunk 3 above)
Comment (facuspagnuolo) [reply mainnet/X-51-attack/TheLobbyGuy.sol]: True!

**Pull request merged**